File: leetcode/125.cs
// 125. valid palindrome

// runtime: beats 68% -- memory: beats 51%
public class Solution {
    public bool IsPalindrome(string s) {    
        int l = 0;
        int r = s.Length -1;

        while(l < r) {
            while (l < r && !char.IsLetterOrDigit(s[l])) l++;
            while (l < r && !char.IsLetterOrDigit(s[r])) r--;

            if (char.ToLower(s[l]) != char.ToLower(s[r])) {
                return false;
            }

            l++;
            r--;
        }

        return true;
    }
}

// runtime: beats 34% -- memory: beats 34%
public class Solution {
    public bool IsPalindrome(string s) {    
        string fix = new string(s.ToLower().Where(char.IsLetterOrDigit).ToArray());
        return fix.SequenceEqual(fix.Reverse());
    }   
}

File: leetcode/125.py
# https://leetcode.com/problems/valid-palindrome/submissions/1953094854/
def isPalindrome(self, s: str) -> bool:
    length = len(s)
    left = 0
    right = length -1

    while left < right:
        print(left, right)

        if not s[left].isalnum():
            left += 1
            continue

        if not s[right].isalnum():
            right -= 1
            continue

        if s[left].lower() != s[right].lower():
            return False


        left += 1
        right -= 1
    
    return True